### ISSUE TITLE
FIXUP: Fix DTBO

### ISSUE THREAD

**Issue by BotchedRPR** [exynos990-mainline/lk3rd#6]
Our lk3rd has code that expects a dtbo without the `mkdtboimg` format from google. This needs to be fixed in order for the device to actually boot Android.

FYI: lib/fdtapi/fdtapi.c has this code.

We need to add support for the dtbo header, as that is what our production devices use

**Comment by halal-beef** [comment]
👍

**Comment by halal-beef** [comment]
Well, 2 major problems have already arose:

- We cannot get the correct board revision most likely cause the voltage tables for our revisions from ADC are either incorrect, or this is the fault of being a third stage BL rather than first stage.

- Doing a test boot with dtbo merging disabled, a modified recovery with complete fdt included, and a force boot to recovery doesn't make it boot, (as discussed this could very well be decon)

**Comment by halal-beef** [comment]
As we saw in our test env, trying to send an already setup fdt has the kernel crash with errors with the clock driver or ACPM

**Comment by halal-beef** [comment]
maybe we are missing some setup that is done before jumping?

**Comment by BotchedRPR** [comment]
> As we saw in our test env, trying to send an already setup fdt has the kernel crash with errors with the clock driver or ACPM

I do believe these are related, the kernel starts doing ACPM stuff on one thread while printing out clock data on the other. Maybe it wants to merge the dtbo in a specific format?

**Comment by halal-beef** [comment]
Clocks could be a read herring, people responsible for trying to get android's 4.19-stable merged into Samsung's kernel had the clock driver crash due to a commit not related to the clock at all

**Comment by halal-beef** [comment]
TODO: Move lk3rd out from 0x90000000, we may be overwriting the Exynos Characteristics Table (ECT) and screwing up fvmap init cause if it. Hopefully if that's the case tomorrow we will have a POC :).

**Comment by halal-beef** [comment]
<0>[    2.802331]  [4:      swapper/0:    1] Kernel panic - not syncing: VFS: Unable to mount root fs on unknown-block(1,0)


SHES ALIVE

**Comment by halal-beef** [comment]
MOSTLY

**Comment by halal-beef** [comment]
https://github.com/user-attachments/assets/f7826ad9-5c95-40d4-b42c-7870c499aeb2

isnt she pretty

**Comment by halal-beef** [comment]
https://github.com/user-attachments/assets/bc57e928-f556-4995-b64b-6d8839653c25

even lineage! (once re creating the lk3rd image to make teegris happy)

**Comment by halal-beef** [comment]
![Image](https://github.com/user-attachments/assets/b715aec9-3ebd-440c-b8fd-21f6d1ce29e7)

big things coming this way?

**Comment by halal-beef** [comment]
Mostly fixed.
DTBO merges and boots to android

TODO:
- passthru cmdline
- passthru some props too
- passthru some reserved mem

**Comment by halal-beef** [comment]
yeah the ADC being at a lower voltage than expected makes me believe that boardrev is not gonna get read correctly

**Comment by halal-beef** [comment]
Every issue has been ironed out, good night
